game-core/src/main/java/games/strategy/engine/lobby/PlayerNameValidation.java

```diff
@@ -0,0 +1,49 @@
+package games.strategy.engine.lobby;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.NoArgsConstructor;
+import org.triplea.lobby.common.LobbyConstants;
+
+/** Utility class to verify player names. */
+@NoArgsConstructor
+public final class PlayerNameValidation {
+  @VisibleForTesting static final int MAX_LENGTH = 40;
+  private static final int MIN_LENGTH = 3;
+
+  public static boolean isValid(final String userName) {
+    return validate(userName) == null;
+  }
+
+  /** Performs username validations intended to be done on client side. */
+  public static String clientSideValidate(final String userName) {
+    return userName.toLowerCase().contains(LobbyConstants.LOBBY_WATCHER_NAME.toLowerCase())
+        ? LobbyConstants.LOBBY_WATCHER_NAME + " cannot be part of a name"
+        : validate(userName);
+  }
+
+  /**
+   * Checks if a username is syntactical valid.
+   *
+   * @return Error message if username is not valid, otherwise returns an error message.
+   */
+  public static String validate(final String userName) {
+    return userName != null
+            && userName.toLowerCase().contains(LobbyConstants.LOBBY_WATCHER_NAME.toLowerCase())
+        ? LobbyConstants.LOBBY_WATCHER_NAME + " cannot be part of a name"
+        : serverSideValidate(userName);
+  }
+
+  /** Less restrictive validation intended to be done on server side. */
+  public static String serverSideValidate(final String userName) {
+    if ((userName == null) || (userName.length() < MIN_LENGTH)) {
+      return "Name is too short (minimum " + MIN_LENGTH + " characters)";
+    } else if (userName.length() > MAX_LENGTH) {
+      return "Name is too long (maximum " + MAX_LENGTH + " characters)";
+    } else if (!userName.matches("[0-9a-zA-Z_-]+")) {
```

Comment:
I know certain unicode characters are problematic, but I don't think we should restrict users to just a subset of ascii. At least allowing common europe characters like äöü é è ê ñ etc. would be nice.

Reply:
Cernel pointed out rendering problems when non-latin characters are entered into chat, we probably should fix that before opening it up to usernames. The test burden is also there to be sure DB can handle it and that we do not get other issues. I wonder if the original motivation for the restriction is to avoid 'leet-speak' names. I like the idea of keeping it simple and not having very fun '|p0VV|\|/\6E', but I think that can also be debated both ways.

I'm afraid to say the update to loosen that up is involved enough it really should not be done as part of this refactoring. The client side validation does need to be done on the server side though as the original code really has a vulnerability where it assumes a person is using an unmodified client, which is a bad assumption. That leads to a situation where you can spoof someone by using a custom client and then adding trailing whitespace to a name or a ' (1)'.

Reply:
Agreed that this should be discussed in another place at another time.

Yes, swing is really stupid when it comes to non-ascii characters sadly. It doesn't even support emoji (I tried), JavaFX should be able to deal with this IIRC ^^